### PR TITLE
Flash decode improvements r5

### PR DESCRIPTION
**Pull request by caixunshiren**
### Ticket
- [x] Deprecate and remove flash decode gpa op
- [x] Support non-causal mode with attention masks in flash decode: #14008

### Checklist
- [x] Post commit CI: https://github.com/tenstorrent/tt-metal/actions/runs/11509420856
- [x] Single Chip Pipeline: https://github.com/tenstorrent/tt-metal/actions/runs/11487582004
- [x] T3k pipeline: https://github.com/tenstorrent/tt-metal/actions/runs/11465687386

### CODE REVIEW

**Review by kpaigwar — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/compute/sdpa_flash_decode.cpp [hunk 1]
```diff
@@ -423,23 +425,25 @@ void MAIN {
...
-    }
-    else {
-        constexpr uint32_t cb_index_id = tt::CB::dataflow0;
-        cb_wait_front(cb_index_id, 1);
-        volatile uint32_t *index_addr_ptr;
-        cb_get_tile(cb_index_id, 0, &index_addr_ptr);
-        cur_pos = index_addr_ptr[4+cur_batch];
-        cb_release_tile(cb_index_id);
-    }
+    uint32_t cur_pos = St*32-1; // default to non-causal, which we do attention on the entire kv cache. In this case we set cur_pos to the last position
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/compute/sdpa_flash_decode.cpp:428]: `constexpr` right-handed side

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/compute/sdpa_flash_decode.cpp [hunk 2]
```diff
@@ -423,23 +425,25 @@ void MAIN {
...
+    if (is_causal) {
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/compute/sdpa_flash_decode.cpp:429]: constexpr

**Review by TT-BrianLiu — APPROVED** [review]
Minor comments

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/dataflow/reader_decode_all.cpp [hunk 3]
```diff
@@ -229,6 +246,29 @@ void kernel_main() {
...
+                        for (uint32_t col = 0; col < Sk_chunk_t; ++col) {
+                            noc_async_read_tile(mask_tile_id, mask_reader, mask_write_ptr);
+                            mask_tile_id++;
+                            mask_write_ptr += mask_tile_bytes;
+
+                            if (++barrier_count == barrier_threshold) {
+                                noc_async_read_barrier();
+                                barrier_count = 0;
+                            }
+                        }
```
Comment (cglagovichTT) [inline ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/dataflow/reader_decode_all.cpp:265]: worthwhile to move this into a function? Looks like the same block exists in both branches

**Review by cglagovichTT — APPROVED** [review]
LGTM

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/dataflow/reader_decode_all.cpp (hunk 3 above)
Comment (caixunshiren) [reply ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/dataflow/reader_decode_all.cpp]: changed into a function call. Good point!

**Review by mtairum — APPROVED** [review]
Looks good. 

@caixunshiren  for llama3 `test_model_quick.py`, have you checked if the PCCs are different? If so please update the tight pcc check on this test 👍

**Review by yieldthought — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

**Pull request merged**